### PR TITLE
feature: HTTP executor can POST multipart

### PR DESCRIPTION
**Pull request by geoffreybauduin**
I need this feature, but I find my method too "hacky" because of encode/decode and behaviors of the used encoding libraries.

**First issue I came up with, in templater:** 
```
templater> Error while marshaling: json: unsupported type: map[interface {}]interface {}
templater> Error while marshaling: json: unsupported type: map[interface {}]interface {}
templater> Error while marshaling: json: unsupported type: map[interface {}]interface {}
```

It appears `json` does not really like `map[interface{}]interface{}` type (and does not know what to do with it), while `yaml` (it's the one producing this type) is okay with it (hence the modification in `templater.go`)

**Second issue I came up with, in the executor**

Basically, `mapstructure.Decode` does not know how to handle the `MultipartForm` field. I had to set values manually (and assert types later)

If you have any ideas on how to make all this less "hacky", please don't hesitate, you probably use more Golang than I do.

Also, I am not sure how you're making sure any change is not breaking the software, so I did not look into testing.

### CODE REVIEW

**Comment by yesnault** [comment]
\i/ Seems good, two points:
 - can you add an example in readme ? (or add a new Readme file like https://github.com/runabove/venom/pull/11/files#diff-79519ffc6c2ccac4e7862584694ee6e0 , but for exec executor) ?
 - can you one-line all err when it's possible ?

cc @fsamin @sguiheux

File: executors/http/executor.go [hunk 1]
```diff
@@ -119,3 +121,71 @@ func (Executor) Run(l *log.Entry, aliases venom.Aliases, step venom.TestStep) (v
...
+	}
+	if e.Body != "" && e.MultipartForm != nil {
+		return nil, fmt.Errorf("Cannot use both 'body' and 'multipart_form'")
+	}
+	var body *bytes.Buffer
+	var writer *multipart.Writer
+	if e.Body != "" {
+		body = bytes.NewBuffer([]byte(e.Body))
+	} else if e.MultipartForm != nil {
+		form := e.MultipartForm.(map[interface{}]interface{})
```
Comment (fsamin) [inline executors/http/executor.go]: you should check if e.MultipartForm can not be cast to (map[interface{}]interface{})

File: executors/http/executor.go [hunk 2]
```diff
@@ -119,3 +121,71 @@ func (Executor) Run(l *log.Entry, aliases venom.Aliases, step venom.TestStep) (v
...
+		body = &bytes.Buffer{}
+		writer = multipart.NewWriter(body)
+		for key_, value_ := range form {
+			key := key_.(string)
```
Comment (fsamin) [inline executors/http/executor.go]: may be you should try to cast e.MultipartForm to (map[string]string) directly to avoid cas of keys and values.
Or you must check `key, ok := key_.(string)`

File: executors/http/executor.go [hunk 3]
```diff
@@ -119,3 +121,71 @@ func (Executor) Run(l *log.Entry, aliases venom.Aliases, step venom.TestStep) (v
...
+func (e Executor) getRequest() (*http.Request, error) {
+	path := fmt.Sprintf("%s%s", e.URL, e.Path)
+	method := e.Method
+	if method == "" {
+		method = "GET"
+	}
+	if e.Body != "" && e.MultipartForm != nil {
+		return nil, fmt.Errorf("Cannot use both 'body' and 'multipart_form'")
+	}
+	var body *bytes.Buffer
```
Comment (fsamin) [inline executors/http/executor.go]: body := &bytes.Buffer{} and remove the else one line 168

File: executors/http/executor.go [hunk 4]
```diff
@@ -119,3 +121,71 @@ func (Executor) Run(l *log.Entry, aliases venom.Aliases, step venom.TestStep) (v
...
+	if e.Body != "" {
+		body = bytes.NewBuffer([]byte(e.Body))
+	} else if e.MultipartForm != nil {
+		form := e.MultipartForm.(map[interface{}]interface{})
+		body = &bytes.Buffer{}
+		writer = multipart.NewWriter(body)
+		for key_, value_ := range form {
+			key := key_.(string)
+			value := value_.(string)
+			if strings.HasPrefix(value, "@") {
```
Comment (fsamin) [inline executors/http/executor.go]: ?

File: executors/http/executor.go [hunk 5]
```diff
@@ -119,3 +121,71 @@ func (Executor) Run(l *log.Entry, aliases venom.Aliases, step venom.TestStep) (v
...
+				// todo: how can we be sure the @ is not the value we wanted to use ?
+				_, err := os.Stat(value[1:])
+				if !os.IsNotExist(err) {
```
Comment (fsamin) [inline executors/http/executor.go]: if _, err := os.Stat(value[1:]); !os.IsNotExist(err) {

File: executors/http/executor.go [hunk 6]
```diff
@@ -119,3 +121,71 @@ func (Executor) Run(l *log.Entry, aliases venom.Aliases, step venom.TestStep) (v
...
+					part, err := writer.CreateFormFile(key, filepath.Base(value[1:]))
+					if err != nil {
+						return nil, err
+					}
+					err = writeFile(part, value[1:])
```
Comment (fsamin) [inline executors/http/executor.go]: one line please

File: executors/http/executor.go [hunk 7]
```diff
@@ -119,3 +121,71 @@ func (Executor) Run(l *log.Entry, aliases venom.Aliases, step venom.TestStep) (v
...
+					if err != nil {
+						return nil, err
+					}
+					continue
+				}
+			}
+			writer.WriteField(key, value)
```
Comment (fsamin) [inline executors/http/executor.go]: missing error checking i guess

Comment (geoffreybauduin) [inline executors/http/executor.go]: You're right, my IDE did not warn me about that. Will fix

File: executors/http/executor.go (hunk 4 above)
Comment (geoffreybauduin) [inline executors/http/executor.go]: I'm getting inspired from curl which loads a file if data is prefixed by "@". Will document that

**Comment by geoffreybauduin** [comment]
I considered all your remarks @fsamin

**Comment by fsamin** [comment]
nice ! thank you very much
